salt/utils/stringutils.py

```diff
@@ -0,0 +1,402 @@
+# -*- coding: utf-8 -*-
+'''
+Functions for manipulating or otherwise processing strings
+'''
+
+# Import Python libs
+from __future__ import absolute_import, print_function, unicode_literals
+import errno
+import fnmatch
+import logging
+import os
+import shlex
+import re
+import time
+
+# Import Salt libs
+from salt.utils.decorators.jinja import jinja_filter
+
+# Import 3rd-party libs
+from salt.ext import six
+from salt.ext.six.moves import range  # pylint: disable=redefined-builtin
+
+log = logging.getLogger(__name__)
+
+
+@jinja_filter('to_bytes')
+def to_bytes(s, encoding=None):
+    '''
+    Given bytes, bytearray, str, or unicode (python 2), return bytes (str for
+    python 2)
+    '''
+    if six.PY3:
+        if isinstance(s, bytes):
+            return s
+        if isinstance(s, bytearray):
+            return bytes(s)
+        if isinstance(s, six.string_types):
+            return s.encode(encoding or __salt_system_encoding__)
+        raise TypeError('expected bytes, bytearray, or str')
+    else:
+        return to_str(s, encoding)
+
+
+def to_str(s, encoding=None):
+    '''
+    Given str, bytes, bytearray, or unicode (py2), return str
+    '''
+    # This shouldn't be six.string_types because if we're on PY2 and we already
+    # have a string, we should just return it.
+    if isinstance(s, str):
+        return s
+    if six.PY3:
+        if isinstance(s, (bytes, bytearray)):
+            # https://docs.python.org/3/howto/unicode.html#the-unicode-type
+            # replace error with U+FFFD, REPLACEMENT CHARACTER
+            return s.decode(encoding or __salt_system_encoding__, "replace")
+        raise TypeError('expected str, bytes, or bytearray not {}'.format(type(s)))
```

Comment:
This is not Python 2.6 compatible code.

Reply:
@isbm why do you say the code at this point is not compatible with Python 2.6?

salt provides six and the code in question, which if unguarded would be python 2.6 incompatible, is guarded by a

if six.PY3:

There is a Python 2.6 incompatibility, but that is in the is_binary() function on line 161 where:

trans_args = ({ord(x): None for x in text_characters},)

is certainly a syntactical problem fro Python 2.6.

@edlane please merge https://github.com/edlane/salt-1/pull/1 into you branch and update this PR ASAP! we need to continue with review and testing.